### DISCUSSION TITLE
Subject: Help Needed with Vehicle Sensors in Autonomous Vehicle Application

### DISCUSSION THREAD

**Discussion by yibork** [eclipse/mosaic#427] [Q&A] [answered]
Hi everyone,

I'm working on an autonomous vehicle simulation and am encountering an issue where the sensors (LIDAR, Radar, etc.) are not being activated correctly in my vehicle application, even though the code appears to be set up properly.

**What I’ve Tried:**
1. I’ve set up the sensors (`LIDAR`, `RADAR_REAR`, `RADAR_FRONT`, `RADAR_LEFT`, `RADAR_RIGHT`) in the `onStartup` method:

```java
   SensorType sensorTypes[] = {SensorType.LIDAR, SensorType.RADAR_REAR, SensorType.RADAR_FRONT, SensorType.RADAR_LEFT, SensorType.RADAR_RIGHT};
   ((VehicleOperatingSystem)this.getOs()).activateVehicleSensors(4.0, sensorTypes);
 ```

   I’m logging the vehicle data in the `onVehicleUpdated` method to check if the sensors are being activated:

 ```java
String sensor = updatedVehicleData.getVehicleSensors().toString();
this.getLog().infoSimTime(this, "Sensor: {}", sensor);
```

**The Issue:**
Despite the above setup, in the logs, all the sensors show as `NOT_EQUIPPED` and the distance is always reported as 0.0. I am getting this log message:

```
2024-10-20 15:36:39,854 INFO  - Sensor: VehicleSensors{distance=DistanceSensor{front=DistanceSensorValue{status=NOT_EQUIPPED distance value=0.0}, back=DistanceSensorValue{status=NOT_EQUIPPED distance value=0.0}, left=DistanceSensorValue{status=NOT_EQUIPPED distance value=0.0}, right=DistanceSensorValue{status=NOT_EQUIPPED distance value=0.0}}, radar=RadarSensor{speedFrontVehicle=-1.0}} (at simulation time 6.000,000,000 s)
```

**What I Need Help With:**
Could someone please help me identify if there's a misconfiguration or missing step to properly activate the sensors in this simulation?
- Is there anything in the VehicleOperatingSystem or VehicleData APIs that I might be overlooking?
- Should I be calling a different method to initialize or activate the sensors correctly?

**Comment by kschrab** [comment, marked as answer]
Hi @yibork,

First, the API is a bit unclear here and also I think we do not cover this aspect in our documentation or tutorials. We apologize for that.

To make thinks a bit more clear: What you can activate via `activateVehicleSensors` depends strongly on the vehicle/traffic simulator used in the background, that is currently SUMO for all users of Eclipse MOSAIC. Since SUMO is a traffic simulator, we cannot expect it to model such vehicle sensors like radar or even LiDAR. For that you usually need a vehicle simulator with a 3D environment, such as Carla or our (commercial) simulator PHABMACS.

However, SUMO provides the distance towards its leading or following vehicle, and we use this information to encode/provide it via the radar sensor. So you can activate `RADAR_FRONT `and `RADAR_REAR `and retrieve the distance towards the leading vehicle, and the following vehicle respectively via `updatedVehicleData.getVehicleSensors()`. For that you need a reasonable range, such as 100m.

```java
SensorType[] sensorTypes = {SensorType.RADAR_FRONT, SensorType.RADAR_REAR};
getOs().activateVehicleSensors(100.0, sensorTypes);
```

```
2024-10-21 09:48:09,575 INFO  - Sensor: VehicleSensors{distance=DistanceSensor{front=DistanceSensorValue{status=VEHICLE_DETECTED distance value=33.49395842672318}, back=DistanceSensorValue{status=VEHICLE_DETECTED distance value=95.54465089665783}, left=DistanceSensorValue{status=NOT_EQUIPPED distance value=0.0}, right=DistanceSensorValue{status=NOT_EQUIPPED distance value=0.0}}, radar=RadarSensor{speedFrontVehicle=36.02312835705467}} (at simulation time 74.000,000,000 s)
```

For internal / performance reasons, you furthermore need to tell our coupling implementation to SUMO that it needs to pull the required  information, by adding the following line to the `sumo_config.json` of your scenario.

```diff
"subscriptions": [ "roadposition", "leader" ]
```

---

tl;dr: You can activate RADAR_FRONT and RADAR_REAR and retrieve the distance towards the leading vehicle, and the following vehicle respectively. For everything else you need a vehicle simulator, such as Carla, for which we cannot open-source a coupling yet.

**Reply by yibork** [reply]
That was really helpful. Thank you so much @kschrab.
